lib/consensus/raft/raft.go

```diff
@@ -0,0 +1,718 @@
+	return false
+}
+
+// WaitForUpdates holds until Raft has synced to the last index in the log
+func (rw *raftWrapper) WaitForUpdates(ctx context.Context) error {
+	//ctx, span := trace.StartSpan(ctx, "consensus/raft/WaitForUpdates")
+	//defer span.End()
+
+	logger.Debug("Raft state is catching up to the latest known version. Please wait...")
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			lai := rw.raft.AppliedIndex()
+			li := rw.raft.LastIndex()
+			logger.Debugf("current Raft index: %d/%d",
+				lai, li)
+			if lai == li {
+				return nil
+			}
+			time.Sleep(waitForUpdatesInterval)
+		}
+	}
+}
+
+func (rw *raftWrapper) WaitForPeer(ctx context.Context, pid string, depart bool) error {
+	//ctx, span := trace.StartSpan(ctx, "consensus/raft/WaitForPeer")
+	//defer span.End()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			peers, err := rw.Peers(ctx)
+			if err != nil {
+				return err
+			}
+
+			if len(peers) == 1 && pid == peers[0] && depart {
+				return errWaitingForSelf
+			}
+
+			found := find(peers, pid)
+
+			// departing
+			if depart && !found {
+				return nil
+			}
+
+			// joining
+			if !depart && found {
+				return nil
+			}
+
+			time.Sleep(50 * time.Millisecond)
+		}
+	}
+}
+
+// Snapshot tells Raft to take a snapshot.
+func (rw *raftWrapper) Snapshot() error {
+	future := rw.raft.Snapshot()
+	err := future.Error()
+	if err != nil && err.Error() != hraft.ErrNothingNewToSnapshot.Error() {
+		return err
+	}
+	return nil
+}
+
+// snapshotOnShutdown attempts to take a snapshot before a shutdown.
+// Snapshotting might fail if the raft applied index is not the last index.
+// This waits for the updates and tries to take a snapshot when the
+// applied index is up to date.
+// It will retry if the snapshot still fails, in case more updates have arrived.
+// If waiting for updates times-out, it will not try anymore, since something
+// is wrong. This is a best-effort solution as there is no way to tell Raft
+// to stop processing entries because we want to take a snapshot before
+// shutting down.
+func (rw *raftWrapper) snapshotOnShutdown() error {
+	var err error
+	for i := 0; i < maxShutdownSnapshotRetries; i++ {
+		ctx, cancel := context.WithTimeout(context.Background(), waitForUpdatesShutdownTimeout)
+		err = rw.WaitForUpdates(ctx)
+		cancel()
+		if err != nil {
+			logger.Warn("timed out waiting for state updates before shutdown. Snapshotting may fail")
+			return rw.Snapshot()
+		}
+
+		err = rw.Snapshot()
+		if err == nil {
+			return nil // things worked
+		}
+
+		// There was an error
+		err = errors.New("could not snapshot raft: " + err.Error())
+		logger.Warnf("retrying to snapshot (%d/%d)...", i+1, maxShutdownSnapshotRetries)
+	}
+	return err
+}
+
+// Shutdown shutdown Raft and closes the BoltDB.
+func (rw *raftWrapper) Shutdown(ctx context.Context) error {
+	//_, span := trace.StartSpan(ctx, "consensus/raft/Shutdown")
+	//defer span.End()
+
+	errMsgs := ""
```

Comment:
I'd use `go.uber.org/multierr.Append / Combine` for this

https://pkg.go.dev/go.uber.org/multierr#hdr-Overview

Reply:
done